### PR TITLE
feat: read bytes rather than read lines when digest to avoid memory issue

### PR DESCRIPTION
**Pull request by dgggIoT**
**Issue #, if available:**
#51 
**Description of changes:**
Rather than using readLines to read the entire line, now we read amount of bytes (default 1KB) for digest. 
If the file does not have enough bytes but contains an entire line, we still digest it.
**Why is this change necessary:**
To avoid memory issue if one line is super large.
**How was this change tested:**
Unit test
**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [x ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/logmanager/model/LogFile.java [hunk 1]
```diff
@@ -50,29 +51,21 @@ public static LogFile[] of(File... files) {
...
-                String oneLine = r.readLine();
-                if (oneLine == null) {
-                    break;
-                }
-                linesRead.add(oneLine);
-            }
+    private byte[] readBytes() {
+        byte[] bytesReadArray = new byte[bytesNeeded];
+        try (InputStream r = Files.newInputStream(this.toPath())) {
+            didRead = r.read(bytesReadArray);
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:60]:
```suggestion
            bytesRead = r.read(bytesReadArray);
```

File: src/main/java/com/aws/greengrass/logmanager/model/LogFile.java [hunk 2]
```diff
@@ -85,16 +78,19 @@ public String hashString() {
             if (!this.exists()) {
                 return fileHash;
             }
-            List<String> lines = getLines();
-            if (lines.size() < linesNeeded) {
-                return fileHash;
+            String bytesReadString = new String(readBytes(), StandardCharsets.UTF_8);
+            // if the we read less then 1KB but it contains an entire line, it is still valid.
+            if (didRead < bytesNeeded && bytesReadString.indexOf('\n') > -1) {
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:83]: Use variable names that represent your intent

```suggestion
    boolean hasLessThanBytesNeeded = didRead < bytesNeeded
    boolean containsEntireLine = bytesReadString.indexOf('\n') > -1
    if (hasLessThanBytesNeeded && containsEntireLine) {
```

Comment (Nelsonochoam) [reply src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:83]: What if the log file just has an empty line?, do we really care about it being a full line? - I think what we should care about is we have read "enough bytes" so we should read hopefully all the ones we require but if there are less than that guarantee that they are more than a minimum

File: src/test/java/com/aws/greengrass/logmanager/LogManagerServiceTest.java [hunk 3]
```diff
@@ -135,13 +137,13 @@ static void setupBefore() throws IOException, InterruptedException {
             assertTrue(file.setWritable(true));
 
             try (OutputStream fileOutputStream = Files.newOutputStream(file.toPath())) {
-                fileOutputStream.write("TEST".getBytes(StandardCharsets.UTF_8));
+                fileOutputStream.write(getRandomTestStrings(bytesNeeded).getBytes(StandardCharsets.UTF_8));
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/LogManagerServiceTest.java:140]: I like this

File: src/test/java/com/aws/greengrass/logmanager/LogManagerServiceTest.java [hunk 4]
```diff
@@ -83,6 +83,8 @@
 import static com.aws.greengrass.logmanager.LogManagerService.SYSTEM_LOGS_COMPONENT_NAME;
 import static com.aws.greengrass.logmanager.LogManagerService.SYSTEM_LOGS_CONFIG_TOPIC_NAME;
 import static com.aws.greengrass.logmanager.LogManagerService.UPLOAD_TO_CW_CONFIG_TOPIC_NAME;
+import static com.aws.greengrass.logmanager.model.LogFile.bytesNeeded;
+import static com.aws.greengrass.logmanager.model.LogFileTest.getRandomTestStrings;
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/LogManagerServiceTest.java:87]: This seems weird importing from another test. Maybe move that method to a LogFileTestUtils class inside the `test` folder

File: src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java [hunk 5]
```diff
@@ -25,53 +28,86 @@ public class LogFileTest {
 
     @TempDir
     static Path directoryPath;
-    private final static int DEFAULT_LINES_FOR_DIGEST_NUM = 1;
+    private final static int DEFAULT_BYTES_FOR_DIGEST_NUM = 1024;
 
-    private void writeFiles(File file, int linesNeeded) throws IOException {
+    public static String getRandomTestStrings(int bytesNeeded) {
+        StringBuilder testStrings = new StringBuilder();
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java:34]: The only bad thing about using this is it doesn't add ned lines hence doesn't simulate a real-ish scenario

File: src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java [hunk 6]
```diff
@@ -25,53 +28,86 @@ public class LogFileTest {
...
             throws IOException, NoSuchAlgorithmException {
         File file = new File(directoryPath.resolve("greengrass_test.log").toUri());
-        writeFiles(file, DEFAULT_LINES_FOR_DIGEST_NUM + 1);
-        StringBuilder msg = new StringBuilder();
-        for (int i = 0; i < DEFAULT_LINES_FOR_DIGEST_NUM; i++) {
-            msg.append("line").append(i + 1);
-        }
+        byte[] bytesArray = getRandomTestStrings(DEFAULT_BYTES_FOR_DIGEST_NUM + 100).getBytes(StandardCharsets.UTF_8);
+        writeFiles(file, bytesArray);
+        LogFile logFile = LogFile.of(file);
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java:93]: Why do this? You could create a LogFile directly since it is just a file. 

```
LogFile file = new LogFile(directoryPath.resolve("greengrass_test.log").toUri());
```

File: src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java [hunk 7]
```diff
@@ -25,53 +28,86 @@ public class LogFileTest {
...
+        String fileHash = logFile.hashString();
+        String msg = new String(Arrays.copyOfRange(bytesArray, 0, DEFAULT_BYTES_FOR_DIGEST_NUM));
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java:95]: Why do ew need to `copyOfRange`?  couldn't we do `new String(bytesArray, StandardCharsets.UTF_8)`

File: src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java [hunk 8]
```diff
@@ -25,53 +28,86 @@ public class LogFileTest {
...
+        assertEquals(fileHash, calculate(msg));
+        file.delete();
+    }
+
+    @Test
+    void GIVEN_log_file_with_less_than_target_lines_but_has_new_line_WHEN_calculate_file_hash_THEN_we_get_hash()
+            throws IOException, NoSuchAlgorithmException {
+        File file = new File(directoryPath.resolve("greengrass_test.log").toUri());
+        StringBuilder testString = new StringBuilder(getRandomTestStrings(DEFAULT_BYTES_FOR_DIGEST_NUM - 100));
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java:104]: Why use a `StringBuilder` here? We could just use the returned string

```java
        StringBuilder testString = getRandomTestStrings(DEFAULT_BYTES_FOR_DIGEST_NUM - 100);
        byte[] bytesArray = testString.getBytes(StandardCharsets.UTF_8);

```

File: src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java [hunk 9]
```diff
@@ -25,53 +28,86 @@ public class LogFileTest {
...
+        testString.append(System.lineSeparator());
+        byte[] bytesArray = testString.toString().getBytes(StandardCharsets.UTF_8);
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java:106]: I just find that reading this makes it hard to follow what you are doing we can make it simple and not be fancy like

```java
StringBuilder builder = new StringBuilder();
builder.append(getRandomTestStrings(DEFAULT_BYTES_FOR_DIGEST_NUM - 100));
builder.append(System.lineSeparator());

writeFile(file, builder.build()) // We can still pass a string here and let the write file convert it to an array
```

File: src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java (hunk 5 above)
Comment (Nelsonochoam) [reply src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java:34]: Edit I saw how you are building the strings below

File: src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java [hunk 10]
```diff
@@ -25,53 +28,86 @@ public class LogFileTest {
 
     @TempDir
     static Path directoryPath;
-    private final static int DEFAULT_LINES_FOR_DIGEST_NUM = 1;
+    private final static int DEFAULT_BYTES_FOR_DIGEST_NUM = 1024;
 
-    private void writeFiles(File file, int linesNeeded) throws IOException {
+    public static String getRandomTestStrings(int bytesNeeded) {
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/model/LogFileTest.java:33]:
```suggestion
    public static String givenAStringOfSize(int bytesNeeded) {
```

File: src/main/java/com/aws/greengrass/logmanager/model/LogFile.java (hunk 1 above)
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:60]: You should handle `didRead` here, rather than having it as a field of this class. I think that makes the API too complicated.

File: src/main/java/com/aws/greengrass/logmanager/model/LogFile.java [hunk 11]
```diff
@@ -85,16 +78,19 @@ public String hashString() {
...
+                fileHash = calculate(bytesReadString.substring(0, bytesReadString.indexOf('\n') + 1));
+            }
+            if (didRead == bytesNeeded) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:86]: use `>=` just because

File: src/main/java/com/aws/greengrass/logmanager/model/LogFile.java (hunk 2 above)
Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:83]: I think there are two reasons:
1. The case of empty line is rare, because it is not a standard way to write logs.  
2. There may be the case that the log file only has content < 1KB, but we should not miss uploading these contents.

Comment (Nelsonochoam) [reply src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:83]: 1) We should handle it and not make that assume it wont happen.

2) Why wouldn't we want to create a hash if we have less bytes than what we want but not a full line? - I think it is better to require at least a minimum instead of expecting a new line

Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:83]: It is hard to define the minimum number. For example, the log file may only has one line, and it only contains 30 bytes. Then the file is rotated. In this example, we still need to upload this line since it is actually an valid log line. But it is hard for us to say 30 is the right number as the minimum number.

File: src/test/java/com/aws/greengrass/logmanager/LogManagerServiceTest.java (hunk 4 above)
Comment (dgggIoT) [reply src/test/java/com/aws/greengrass/logmanager/LogManagerServiceTest.java:87]: Probably keep it for now, if we need more such utils method, then we can create such class?

File: src/main/java/com/aws/greengrass/logmanager/model/LogFile.java (hunk 2 above)
Comment (Nelsonochoam) [reply src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:83]: I think we are mixing 2 separate things here:

1) The how many bytes we need to compute a hash has nothing to do with the fact that the file might have a single line or multiple lines. To address your example we might have a file with a single line and not line break that is below the bytes you require that gets rotated and we still would have to upload it.

2) The other problem with this is that calculating the hash with this implementation is not deterministic - Depending on when you call `hashString` it might return different values
     2.1) If you call it and you have less than what you require it will return a value
     2.2) If the file then get more lines written to it and you can get all the bytes you need you will get a different hash.


We either have to memoize the hash value after you first compute it or have a firm required number of bytes you need.

Comment (dgggIoT) [reply src/main/java/com/aws/greengrass/logmanager/model/LogFile.java:83]: Agree. I will change it to the following:
1. If there is a new line before 1KB, then we take a new line
2. Otherwise, we read 1KB to prevent crazy long one line message.

**Review by MikeDombo — APPROVED** [review]
